internal/datastore/proxy/relationshipintegrity.go

```diff
@@ -0,0 +1,397 @@
+package proxy
+
+import (
+	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"fmt"
+	"hash"
+	"strings"
+	"sync"
+	"time"
+
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/authzed/spicedb/pkg/datastore"
+	"github.com/authzed/spicedb/pkg/datastore/options"
+	corev1 "github.com/authzed/spicedb/pkg/proto/core/v1"
+	"github.com/authzed/spicedb/pkg/spiceerrors"
+	"github.com/authzed/spicedb/pkg/tuple"
+)
+
+// KeyConfig is a configuration for a key used to sign relationships.
+type KeyConfig struct {
+	ID        string
+	ExpiredAt *time.Time
+	Bytes     []byte
+}
+
+type hmacConfig struct {
+	keyID     string
+	expiredAt *time.Time
+	hmacPool  *sync.Pool
+}
+
+var (
+	alg         = sha256.New
+	versionByte = byte(0x01)
+)
+
+// NewRelationshipIntegrityProxy creates a new datastore proxy that ensures the integrity of
+// relationships by using HMACs to sign the data. The current key is used to sign new data,
+// and the expired keys are used to verify old data, if any.
+func NewRelationshipIntegrityProxy(ds datastore.Datastore, currentKey KeyConfig, expiredKeys []KeyConfig) (datastore.Datastore, error) {
+	currentKeyHMAC := hmacConfig{
+		keyID:     currentKey.ID,
+		expiredAt: currentKey.ExpiredAt,
+		hmacPool: &sync.Pool{
```

Comment:
I'm sure it depends, but for tiger I found that a buffered channel performed better than sync.Pool (`select` on retrieving from the pool to return a new instance if the pool is empty, similar to sync.Pool)